plugins/tedge_apt_plugin/src/lib.rs

```diff
@@ -103,29 +117,67 @@ fn run_op(apt: AptCli) -> Result<ExitStatus, InternalError> {
+                    )
+                    .as_str(),
+                ) {
+                    Ok(filter) => Some(filter),
+                    Err(err) => {
+                        eprintln!("tedge-apt-plugin fails to list packages with matching name and maintainer: {err}");
+                        std::process::exit(1)
+                    }
+                },
+            };
+
+            let exclude_filter = match (&exclude_name, &exclude_maintainer) {
+                (None, None) => None,
 
                 _ => match Regex::new(
                     format!(
-                        r"(^{}\t.*|^\S+\t\S+\t{}\s+.*)install ok installed",
-                        name.unwrap_or_default(),
-                        maintainer.unwrap_or_default()
+                        r"(^{}\t.*|^\S+\t\S+\t{}\s+.*)",
+                        exclude_name.unwrap_or_default(),
+                        exclude_maintainer.unwrap_or_default()
                     )
                     .as_str(),
                 ) {
-                    Ok(filter) => filter,
+                    Ok(filter) => Some(filter),
                     Err(err) => {
                         eprintln!("tedge-apt-plugin fails to list packages with matching name and maintainer: {err}");
                         std::process::exit(1)
                     }
                 },
             };
 
-            for line in stdout.trim_end().lines() {
-                if filter.is_match(line) {
-                    let (name, version) = get_name_and_version(line);
-                    println!("{name}\t{version}");
+            match (include_filter.as_ref(), exclude_filter.as_ref()) {
+                (Some(include_filter), Some(exclude_filter)) => {
+                    let installed_filter = Regex::new(r"install ok installed").unwrap();
+                    for line in stdout.trim_end().lines() {
+                        if installed_filter.is_match(line)
+                            && (include_filter.is_match(line) || !exclude_filter.is_match(line))
+                        {
+                            let (name, version) = get_name_and_version(line);
+                            println!("{name}\t{version}");
+                        }
+                    }
+                }
+                _ => {
+                    let include_filter = include_filter
+                        .unwrap_or_else(|| Regex::new(r"install ok installed").unwrap());
+                    let exclude_filter =
+                        exclude_filter.unwrap_or_else(|| Regex::new(r"^$").unwrap());
+                    for line in stdout.trim_end().lines() {
+                        if include_filter.is_match(line) && !exclude_filter.is_match(line) {
+                            let (name, version) = get_name_and_version(line);
+                            println!("{name}\t{version}");
+                        }
+                    }
                 }
```

Comment:
What if an *include* filter is given but no *exclude* filter? Are only the *installed* packages correctly listed?
Removing case matching would really help to be confident.

```rust
let installed_filter = Regex::new(r"install ok installed").unwrap();
let include_filter = include_filter.unwrap_or(Regex::new(r".*").unwrap());
let exclude_filter = exclude_filterunwrap_or(Regex::new(r^$").unwrap());
for line in stdout.trim_end().lines() {
       if installed_filter.is_match(line)
             && (include_filter.is_match(line) || !exclude_filter.is_match(line))
       {
               let (name, version) = get_name_and_version(line);
               println!("{name}\t{version}");
        }
}
```

If you dislike this suggestion, I would prefer the very first proposal you published: i.e. with the 4 cases being explicit. At least, it's easy to check correctness, while this is not the case with this intermediate solution.

Reply:
I still think it is still better than four cases. I forgot that check for installed packages should also be added in include_filter: 
```rust
Regex::new(format!(r"(^{}\t.*|^\S+\t\S+\t{}\s+.*installed ok install)", include_name.unwrap_or_default(), include_maintainer.unwrap_or_default())
```
Anyway, I like the proposed solution more, it is indeed easier to verify.